### ISSUE TITLE
postgres cluster replicas not in sync with master

### ISSUE THREAD

**Issue by haroonb** [zalando/postgres-operator#949]
Version 1.4.0

Hello we ran into a major issue today. 

From the beginning we switched from another postgres-cluster to the operator managed postgres-cluster. We switched with the following steps.

1. deploy the operator managed cluster with 3 replicas
2. dumpall old cluster
3. scale the operator managed cluster to 1 replica, to avoid master switch during restore
4. scaled the operator managed cluster back to 3 replicas.

At this point everything was working fine and we assumed, that the secondary pods would sync with the master. We did not verify this assumption, that was a mistake on our part.

Yesterday evening we updated the configuration of the postgres cluster with `- hostssl replication user x.x.x.x/x md5` to support pg_basebackup. As expected the cluster did a rolling update and applied the updates. During that rolling updated occured a master switch, as expected. The fatal thing was, that the new leader had a empty database, meaning that the pg_basebackup we have of that day is empty, no user could login in the morning and all out applications weren't working.

We where in luck, that the pod-0 was in sync with current data, so we scaled to 1 replica and got everything working. To see if the cluster would recover on it's own we scaled back to 3 and saw, that pod-1 wasn't syncing and the pod-2 was stuck. As a last resort we scaled back to 1 replica deleted the pv,pvc of pod-1/2 and scaled back to 3 replicas. Fortunately both pods started the bootstrap and are now in sync with the leader.

The only log I still have is of pod-2 which would start the database and the operator. I hope to figure what went wrong and how to avoid it in the future. I hope anyone could assist with analysing why the new leader had a empty database.

```
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52,234 WARNING: Postgresql is not running.
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52,234 INFO: Lock owner: acid-postgres-cluster-0; I am acid-postgres-cluster-2
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52,243 INFO: pg_controldata:
acid-postgres-cluster-2 postgres   pg_control version number: 1201
acid-postgres-cluster-2 postgres   Catalog version number: 201909212
acid-postgres-cluster-2 postgres   Database system identifier: 6818416449794248770
acid-postgres-cluster-2 postgres   Database cluster state: shut down in recovery
acid-postgres-cluster-2 postgres   pg_control last modified: Wed Apr 29 07:31:45 2020
acid-postgres-cluster-2 postgres   Latest checkpoint location: 1/C4000028
acid-postgres-cluster-2 postgres   Latest checkpoint's REDO location: 1/C4000028
acid-postgres-cluster-2 postgres   Latest checkpoint's REDO WAL file: 0000000100000001000000C4
acid-postgres-cluster-2 postgres   Latest checkpoint's TimeLineID: 1
acid-postgres-cluster-2 postgres   Latest checkpoint's PrevTimeLineID: 1
acid-postgres-cluster-2 postgres   Latest checkpoint's full_page_writes: on
acid-postgres-cluster-2 postgres   Latest checkpoint's NextXID: 0:162485
acid-postgres-cluster-2 postgres   Latest checkpoint's NextOID: 390424
acid-postgres-cluster-2 postgres   Latest checkpoint's NextMultiXactId: 1182
acid-postgres-cluster-2 postgres   Latest checkpoint's NextMultiOffset: 2572
acid-postgres-cluster-2 postgres   Latest checkpoint's oldestXID: 480
acid-postgres-cluster-2 postgres   Latest checkpoint's oldestXID's DB: 1
acid-postgres-cluster-2 postgres   Latest checkpoint's oldestActiveXID: 0
acid-postgres-cluster-2 postgres   Latest checkpoint's oldestMultiXid: 1
acid-postgres-cluster-2 postgres   Latest checkpoint's oldestMulti's DB: 1
acid-postgres-cluster-2 postgres   Latest checkpoint's oldestCommitTsXid: 0
acid-postgres-cluster-2 postgres   Latest checkpoint's newestCommitTsXid: 0
acid-postgres-cluster-2 postgres   Time of latest checkpoint: Tue Apr 28 16:58:51 2020
acid-postgres-cluster-2 postgres   Fake LSN counter for unlogged rels: 0/3E8
acid-postgres-cluster-2 postgres   Minimum recovery ending location: 1/C40000A0
acid-postgres-cluster-2 postgres   Min recovery ending loc's timeline: 1
acid-postgres-cluster-2 postgres   Backup start location: 0/0
acid-postgres-cluster-2 postgres   Backup end location: 0/0
acid-postgres-cluster-2 postgres   End-of-backup record required: no
acid-postgres-cluster-2 postgres   wal_level setting: replica
acid-postgres-cluster-2 postgres   wal_log_hints setting: on
acid-postgres-cluster-2 postgres   max_connections setting: 300
acid-postgres-cluster-2 postgres   max_worker_processes setting: 8
acid-postgres-cluster-2 postgres   max_wal_senders setting: 10
acid-postgres-cluster-2 postgres   max_prepared_xacts setting: 0
acid-postgres-cluster-2 postgres   max_locks_per_xact setting: 64
acid-postgres-cluster-2 postgres   track_commit_timestamp setting: off
acid-postgres-cluster-2 postgres   Maximum data alignment: 8
acid-postgres-cluster-2 postgres   Database block size: 8192
acid-postgres-cluster-2 postgres   Blocks per segment of large relation: 131072
acid-postgres-cluster-2 postgres   WAL block size: 8192
acid-postgres-cluster-2 postgres   Bytes per WAL segment: 16777216
acid-postgres-cluster-2 postgres   Maximum length of identifiers: 64
acid-postgres-cluster-2 postgres   Maximum columns in an index: 32
acid-postgres-cluster-2 postgres   Maximum size of a TOAST chunk: 1996
acid-postgres-cluster-2 postgres   Size of a large-object chunk: 2048
acid-postgres-cluster-2 postgres   Date/time type storage: 64-bit integers
acid-postgres-cluster-2 postgres   Float4 argument passing: by value
acid-postgres-cluster-2 postgres   Float8 argument passing: by value
acid-postgres-cluster-2 postgres   Data page checksum version: 1
acid-postgres-cluster-2 postgres   Mock authentication nonce: 7871782b6c6c26ae41e3435e066c04c8c83b60b63ef8b7b62c3045fc16d1afec
acid-postgres-cluster-2 postgres
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52,244 INFO: Lock owner: acid-postgres-cluster-0; I am acid-postgres-cluster-2
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52,245 INFO: Lock owner: acid-postgres-cluster-0; I am acid-postgres-cluster-2
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52,246 INFO: starting as a secondary
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52 UTC [519]: [1-1] 5ea92f0c.207 0     LOG:  Auto detecting pg_stat_kcache.linux_hz parameter...
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52 UTC [519]: [2-1] 5ea92f0c.207 0     LOG:  pg_stat_kcache.linux_hz is set to 500000
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52 UTC [519]: [3-1] 5ea92f0c.207 0     LOG:  starting PostgreSQL 12.2 (Ubuntu 12.2-1.pgdg18.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0, 64-bit
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52 UTC [519]: [4-1] 5ea92f0c.207 0     LOG:  listening on IPv4 address "0.0.0.0", port 5432
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52 UTC [519]: [5-1] 5ea92f0c.207 0     LOG:  listening on IPv6 address "::", port 5432
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52 UTC [519]: [6-1] 5ea92f0c.207 0     LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52,583 INFO: postmaster pid=519
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52 UTC [519]: [7-1] 5ea92f0c.207 0     LOG:  redirecting log output to logging collector process
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52 UTC [519]: [8-1] 5ea92f0c.207 0     HINT:  Future log output will appear in directory "../pg_log".
acid-postgres-cluster-2 postgres /var/run/postgresql:5432 - rejecting connections
acid-postgres-cluster-2 postgres /var/run/postgresql:5432 - no response
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52,723 INFO: Lock owner: acid-postgres-cluster-0; I am acid-postgres-cluster-2
acid-postgres-cluster-2 postgres 2020-04-29 07:38:52,723 INFO: failed to start postgres

```

**Comment by FxKu** [comment]
@haroonb how big is this cluster? And how time passed until the master was ready and you changed the configuration. Sounds odd to me that a volume on a former replica is empty. Maybe @CyberDem0n can clearify.

Could be that the replica pods were still catching up when the rolling update happen. For the new `v1.5.0` release we made such updates more robust by checking the health of all replicas (calling Patroni for the replica status).

**Comment by haroonb** [comment]
@FxKu This all happened in a period of about a week.
Day 1 we shut the old cluster down and switched to the operator managed cluster. The cluster size is 3. After that the cluster was in testing for about 4 days, when the configuration was updated. The size of the restored database wasn't that big either, only about 13 GB.

I have tried  to recreate this issue, but without success.

**Comment by haroonb** [comment]
Hi, today I managed to recreate this issue. I now have a master and 2 slaves which are out of sync. What logs/Information can I provide to assist?

**Comment by FxKu** [comment]
@haroonb out of sync by what means? Is there a WAL delay (can be checked with `patronictl list` on one of the pods)? Does the replication streaming work, in general? What do the pod logs show during the bootstrap phase (create a Github gist, don't post it here)?

**Comment by haroonb** [comment]
@FxKu There seems to be a WAL delay of 2668 MB. I created a gist with the relevant informations. Let me know if I can assist any further.

https://gist.github.com/haroonb/09460dd9e056c70574eaa9a3ea7cf43e

**Comment by FxKu** [comment]
Hm, not so easy to tell. I would have thought to just give your pods more resources so that the replicas don't fall behind.

I can see one error about a replica endpoint already existing, although the logs say that it could be found before. Strange. Does every SYNC event have this error?

**Comment by haroonb** [comment]
@FxKu Hello, sorry for the late reply, I wasn't in the office for a week. 

I don't know which SYNC events you're refering too. I have attached a updated log of the current operator status. I can't attach the cluster log, because its too long, but there is no update, the cluster is still out of sync.

https://gist.github.com/haroonb/43ec4fa02902cd8af2166a771c021a31

**Comment by FxKu** [comment]
@haroonb that's what I meant.
So there's still a replica lag? Does it stay constant? Is it growing? The value you've mentioned last time is not sooo much to not keep up.

What K8s stack are you using? Maybe you have too many IOPS for the chosen instance type.

Operator logs look fine. Seems like a general Postgres maintenance issue and less operator related. That we can not fix for you here.

**Comment by haroonb** [comment]
> @haroonb that's what I meant.
> So there's still a replica lag? Does it stay constant? Is it growing? The value you've mentioned last time is not sooo much to not keep up.

@FxKu the cluster in not beeing used, I only created it for the purpose to recreate this issue. There is not any new data beeing generated. I wanted to keep it as is, so we could get to the bottom of this.
> 
> What K8s stack are you using? Maybe you have too many IOPS for the chosen instance type.
k8s on docker. The cluster is idle, since I managed to recreate this issue.

> Operator logs look fine. Seems like a general Postgres maintenance issue and less operator related. That we can not fix for you here.
I didn't do anything specific with the cluster. It's a fresh install and the only action from my part was scaling the cluster to 1 and restoring a backup. I can provide the yaml files, if that would assist in finding the cause of this.

**Comment by haroonb** [comment]
I repaired the cluster today, because I needed to test some other things and needed slaves with data in them.

The fix was very straight forward. I just scaled the statefulset to 1 replica and deleted the pvc/pv and scaled back to 3 replicas. As soon as the pods came back up they started to sync.

I'll attach another log of that https://gist.github.com/haroonb/f0a2289cafb53b7ff8d76ab8c532dcd7.

Considering the outcome and the way I fixed the cluster I don't think it was a administration/configuration issue.
